Floors: Dynamic fetch

pm-nikhil-vaidya opened the pull request:
Design document of dynamic fetch mechanism
https://docs.google.com/document/d/1Lb302nt6nDYQxeflzISjtIWpE6ifExdoT8PSPm8YSKk/edit#heading=h.bxed3ptmj442

Issue + Specs: https://github.com/prebid/prebid-server/issues/1162

AlexBVolcy commented on config/config.go:
```diff
@@ -1020,8 +1026,16 @@ func SetupViper(v *viper.Viper, filename string, bidderInfos BidderInfos) {
+	v.SetDefault("account_defaults.price_floors.fetch.max_age_sec", 600)
+
+	//Defaults for Price floor fetcher
+	v.SetDefault("price_floor.fetcher.worker", 20)
+	v.SetDefault("price_floor.fetcher.capacity", 20000)
```
How were these default values decided? I couldn't find them in the floors documentation

pm-nikhil-vaidya replied:
20000 will be large number if prebid is hosted by a publisher, as he may have maximum 2-3 urls, but if we think about hosting company who are working with multiple company then this number will be valid in maximum case
We can make this configurable, through configuration file.

AlexBVolcy commented on exchange/exchange.go:
```diff
@@ -125,7 +126,7 @@ func (randomDeduplicateBidBooleanGenerator) Generate() bool {
 }
 
-func NewExchange(adapters map[openrtb_ext.BidderName]AdaptedBidder, cache prebid_cache_client.Client, cfg *config.Configuration, syncersByBidder map[string]usersync.Syncer, metricsEngine metrics.MetricsEngine, infos config.BidderInfos, gdprPermsBuilder gdpr.PermissionsBuilder, currencyConverter *currency.RateConverter, categoriesFetcher stored_requests.CategoryFetcher, adsCertSigner adscert.Signe ...
+func NewExchange(adapters map[openrtb_ext.BidderName]AdaptedBidder, cache prebid_cache_client.Client, cfg *config.Configuration, syncersByBidder map[string]usersync.Syncer, metricsEngine metrics.MetricsEngine, infos config.BidderInfos, gdprPermsBuilder gdpr.PermissionsBuilder, currencyConverter *currency.RateConverter, categoriesFetcher stored_requests.CategoryFetcher, adsCertSigner adscert.Signe ...
```
Nitpick: For camel case I think this variable should be named `priceFloorFetcher`

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+type PriceFloorFetcher struct {
+	pool            WorkerPool      // Goroutines worker pool
+	fetchQueue      FetchQueue      // Priority Queue to fetch floor data
+	fetchInprogress map[string]bool // Map of URL with fetch status
```
Nitpick: For camel case i think this variable should be named `fetchInProgress`

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+func (f *PriceFloorFetcher) Fetch(configs config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
+
```
Nitpick: Please remove this blank line

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+	cacheExpiry     time.Duration   // cache expiry time
+}
+
+func (f *PriceFloorFetcher) Fetch(configs config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
```
Nitpick: Could you re-name `configs` to be `config`?

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+	// Check for floors JSON in cache
+	result, found := f.Get(configs.Fetch.URL)
+	if found {
+		fetcheRes, ok := result.(*openrtb_ext.PriceFloorRules)
```
Could you rename `fetcheRes` to be `fetcherResult`

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+func NewPriceFloorFetcher(maxWorkers, maxCapacity, cacheCleanUpInt, cacheExpiry int) *PriceFloorFetcher {
+
```
Could you remove blank line

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+	floorFetcher := PriceFloorFetcher{
+		pool:            pond.New(maxWorkers, maxCapacity, pond.PanicHandler(workerPanicHandler)),
+		fetchQueue:      make(FetchQueue, 0, 100),
+		fetchInprogress: make(map[string]bool),
```
`fetchInprogress` --> `fetchInProgress`

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+func (f *PriceFloorFetcher) worker(configs config.AccountFloorFetch) {
+
```
Could you remove blank line

pm-nikhil-vaidya replied:
done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+func (f *PriceFloorFetcher) Fetcher() {
+
```
Could you remove blank line

AlexBVolcy commented on floors/floors.go:
```diff
@@ -209,3 +220,98 @@ func updateFloorsInRequest(bidRequestWrapper *openrtb_ext.RequestWrapper, priceF
+		}
+	}
+
+	// Question :: What should be the behaviour if request floorMin present and currency is not defined?
```
Has this question been answered yet? If not, can you add it as a comment on the PR?

pm-nikhil-vaidya replied:
Added in the comments

pm-nikhil-vaidya replied:
Resolved

AlexBVolcy commented on floors/floors.go:
```diff
@@ -209,3 +220,98 @@ func updateFloorsInRequest(bidRequestWrapper *openrtb_ext.RequestWrapper, priceF
+		enforceRate = reqFloors.Enforcement.EnforceRate
+	}
+
+	// Question :: Should we copy entire Enforcement Object from request?
```
Has this question been answered yet? If not, can you add it as a comment on the PR?

AlexBVolcy commented on router/router.go:
```diff
@@ -213,8 +214,12 @@ func New(cfg *config.Configuration, rateConvertor *currency.RateConverter) (r *R
 		glog.Fatalf("Failed to create ads cert signer: %v", err)
 	}
 
+	//Price Floor Fetcher
```
I think this comment can be deleted

pm-nikhil-vaidya replied:
Done

AlexBVolcy reviewed (commented):
I left this comment a few times, but saw it in many other places, but could you remove any blank lines from the beginning of functions?

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+
+func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
+	if !config.UseDynamicData || len(config.Fetch.URL) == 0 || !validator.IsURL(config.Fetch.URL) {
+		return nil, openrtb_ext.FetchNone
```
Is there a way to cover this line with a test?

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+}
+
+func workerPanicHandler(p interface{}) {
+	glog.Errorf("floor fetcher worker panicked: %v", p)
```
Can this error line be covered by tests

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/floors.go:
```diff
@@ -209,3 +219,97 @@ func updateFloorsInRequest(bidRequestWrapper *openrtb_ext.RequestWrapper, priceF
+func mergeFloors(reqFloors *openrtb_ext.PriceFloorRules, fetchFloors *openrtb_ext.PriceFloorRules, conversions currency.Conversions) *openrtb_ext.PriceFloorRules {
+	mergedFloors := fetchFloors.DeepCopy()
+	if mergedFloors.Enabled == nil {
+		mergedFloors.Enabled = new(bool)
```
Can you cover this line with a test?

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on openrtb_ext/floors.go:
```diff
@@ -145,3 +151,59 @@ type ExtImp struct {
+	}
+
+	newEnforcement := *enforcement
+	newEnforcement.EnforceJS = ptrutil.Clone(enforcement.EnforceJS)
+	newEnforcement.EnforcePBS = ptrutil.Clone(enforcement.EnforcePBS)
+	newEnforcement.FloorDeals = ptrutil.Clone(enforcement.FloorDeals)
+	newEnforcement.BidAdjustment = ptrutil.Clone(enforcement.BidAdjustment)
+
+	return &newEnforcement
```
These lines aren't covered by tests. Could you update this?

pm-nikhil-vaidya replied:
Done

AlexBVolcy reviewed (commented):
This looks great! Close to approving, some comments about test coverage. Could you also merge with master to fix the merge conflicts? Thanks!

AlexBVolcy commented:
This looks good! I will wait for @bsardo's review before approving.

SyntaxNode commented on config/account.go:
```diff
@@ -58,17 +58,29 @@ type AccountCCPA struct {
 }
 
-func (pf *AccountPriceFloors) validate(errs []error) []error {
+// Dynamic price floors fetching configuration
```
Please follow Go standards for all comments, where the comment begins with the type name. For example:

```
// AccountFloorFetch defines the configuration for dynamic floors fetching.
```

pm-nikhil-vaidya replied:
Done

SyntaxNode commented on config/account.go:
```diff
@@ -58,17 +58,29 @@ type AccountCCPA struct {
+	UseDynamicData         bool              `mapstructure:"use_dynamic_data" json:"use_dynamic_data"`
+	MaxRule                int               `mapstructure:"max_rules" json:"max_rules"`
+	MaxSchemaDims          int               `mapstructure:"max_schema_dims" json:"max_schema_dims"`
+	Fetch                  AccountFloorFetch `mapstructure:"fetch" json:"fetch"`
```
Nitpick: Consider calling this a `Fetcher` to keep in line with current and future PBS terminology.

pm-nikhil-vaidya replied:
Done

SyntaxNode commented on config/account.go:
```diff
@@ -81,6 +93,30 @@ func (pf *AccountPriceFloors) validate(errs []error) []error {
+
+	if !(pf.Fetch.MaxAge >= 600 && pf.Fetch.MaxAge < math.MaxInt32) {
+		errs = append(errs, fmt.Errorf(`account_defaults.price_floors.fetch.max_age_sec should not be less than 600 seconds and greater than maximum integer value`))
+	}
```
We are we opinionated about enforcing a minimum value?

pm-nikhil-vaidya replied:
Value validations
enforce-floors-rate >=0 and <= 100
fetch.period-sec must be <= fetch.max-age-sec
fetch.period-sec >= 300 (cannot poll more often than every 5 min)
fetch.max-age-sec >= 600 (must allow data to be valid for at least 10 min)
fetch.max-age-sec < MAXINT
fetch.timeout-ms must be > 10 and < 10,000
fetch.max-rules >= 0 and < MAXINT
fetch.max-file-size-kb >= 0 and < MAXINT
fetch.max-schema-dims >= 0 and < 20
max-rules >= 0 and < MAXINT
max-schema-dims >= 0 and < 20

pm-nikhil-vaidya replied:
From The PRD

SyntaxNode commented on config/account.go:
```diff
@@ -81,6 +93,30 @@ func (pf *AccountPriceFloors) validate(errs []error) []error {
+	}
+
+	if !(pf.Fetch.Timeout > 10 && pf.Fetch.Timeout < 10000) {
+		errs = append(errs, fmt.Errorf(`account_defaults.price_floors.fetch.timeout_ms should be between 10 to 10,000 mili seconds`))
```
Nitpick: "mili seconds" -> "milliseconds"

pm-nikhil-vaidya replied:
Done

SyntaxNode commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+	floorData, fetchedMaxAge := fetchAndValidate(configs)
+	if floorData != nil {
+		// Update cache with new floor rules
+		glog.Infof("Updating Value in cache for URL %s", configs.URL)
```
Could be useful for debugging. For production, consider adding a per-account metric for monitoring the floors fetcher.

pm-nikhil-vaidya replied:
Yes adding metric to this will be next task, will add them soon

SyntaxNode commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+	}
+
+	// Send to refetch channel
+	f.configReceiver <- FetchInfo{AccountFloorFetch: configs, FetchTime: time.Now().Add(time.Duration(configs.Period) * time.Second).Unix(), RefetchRequest: true}
```
For testing, please extract all calls to `time.Now()` with our existing `timeutil.Time` interface.

pm-nikhil-vaidya replied:
Done

SyntaxNode commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+		return nil, 0, errors.New("error while forming http fetch request : " + err.Error())
+	}
+
+	httpResp, err := http.DefaultClient.Do(httpReq)
```
Please do not use the default client. There is not much control available through this connection pool. I recommend you create a new floors fetcher specific pool and pass it down from `router.go`.

pm-nikhil-vaidya replied:
Should we also declare transport? or it will be okay if we use default transport?

SyntaxNode commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+	"github.com/alitto/pond"
+	validator "github.com/asaskevich/govalidator"
+	"github.com/golang/glog"
+	"github.com/patrickmn/go-cache"
```
Could we use `freecache` as we do with the other fetchers?

pm-nikhil-vaidya replied:
Done

pm-nikhil-vaidya commented on config/account.go:
```diff
@@ -58,17 +58,29 @@ type AccountCCPA struct {
+	MaxRules    int    `mapstructure:"max_rules" json:"max_rules"`
+	MaxAge      int    `mapstructure:"max_age_sec" json:"max_age_sec"`
+	Period      int    `mapstructure:"period_sec" json:"period_sec"`
+	AccountID   string `mapstructure:"accountID" json:"accountID"`
```
This was added for the metric purpose in future, if fetcher is not working properly then we should log the metric for that particular account, we will be adding metric into this

pm-nikhil-vaidya commented on config/config.go:
```diff
@@ -1020,8 +1026,16 @@ func SetupViper(v *viper.Viper, filename string, bidderInfos BidderInfos) {
+	v.SetDefault("account_defaults.price_floors.fetch.max_age_sec", 600)
+
+	//Defaults for Price floor fetcher
+	v.SetDefault("price_floor.fetcher.worker", 20)
+	v.SetDefault("price_floor.fetcher.capacity", 20000)
```
Okay sure, what is the ideal place for these configuration in the config file

pm-nikhil-vaidya replied:
I already made part of configuration
You can refer these lines
https://github.com/prebid/prebid-server/pull/2732/files#r1269224255
let me know if this is correct place or need to move somewhere else

pm-nikhil-vaidya replied:
We made this configurable keeping in mind that whoever using the prebid may have different requirement
So if anyone is working with 2-3 urls then it doesn't need that many worker and buffer
So if you want to hard code it then i will make it.
Let me know your thoughts

pm-nikhil-vaidya commented on config/account.go:
```diff
@@ -58,17 +58,29 @@ type AccountCCPA struct {
+	Timeout     int    `mapstructure:"timeout_ms" json:"timeout_ms"`
+	MaxFileSize int    `mapstructure:"max_file_size_kb" json:"max_file_size_kb"`
+	MaxRules    int    `mapstructure:"max_rules" json:"max_rules"`
+	MaxAge      int    `mapstructure:"max_age_sec" json:"max_age_sec"`
```
Done

pm-nikhil-vaidya commented on config/config.go:
```diff
@@ -1020,8 +1026,16 @@ func SetupViper(v *viper.Viper, filename string, bidderInfos BidderInfos) {
 	v.SetDefault("account_defaults.price_floors.max_rules", 100)
 	v.SetDefault("account_defaults.price_floors.max_schema_dims", 3)
-	v.SetDefault("account_defaults.events_enabled", false)
+	v.SetDefault("account_defaults.price_floors.fetch.enabled", false)
+	v.SetDefault("account_defaults.price_floors.fetch.timeout_ms", 100)
+	v.SetDefault("account_defaults.price_floors.fetch.period_sec", 300)
+	v.SetDefault("account_defaults.price_floors.fetch.max_age_sec", 600)
```
Done

pm-nikhil-vaidya commented on config/config.go:
```diff
@@ -109,7 +109,13 @@ type Configuration struct {
+}
+
+type PriceFloorFetcher struct {
+	Worker   int `mapstructure:"worker"`
```
Fetcher Configuration

pm-nikhil-vaidya commented on config/config.go:
```diff
@@ -109,7 +109,13 @@ type Configuration struct {
 
 type PriceFloors struct {
-	Enabled bool `mapstructure:"enabled"`
+	Enabled bool              `mapstructure:"enabled"`
+	Fetcher PriceFloorFetcher `mapstructure:"fetcher"`
+}
+
+type PriceFloorFetcher struct {
+	Worker   int `mapstructure:"worker"`
+	Capacity int `mapstructure:"capacity"`
 }
```
Fetcher configuration

pm-nikhil-vaidya commented on config/config_test.go:
```diff
@@ -477,6 +484,11 @@ account_defaults:
         max_schema_dims: 5
+        fetch:
+          enabled: true
+          timeout_ms: 100
+          period_sec: 300
+          max_age_sec: 600
```
Done

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+var refetchCheckInterval = 300
```
This is very specific to price floor implementation, after this interval, fetcher will check if there are any urls available to fetch. Its internal implementation value.
In order to configure this, it will require fetcher implementation knowledge, we can make it configurable
Let me know your thoughts where we can keep the configuration and what is the minimum and maximum value

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+
+func NewPriceFloorFetcher(maxWorkers, maxCapacity, cacheCleanUpInt, cacheExpiry int) *PriceFloorFetcher {
+	floorFetcher := PriceFloorFetcher{
+		pool:            pond.New(maxWorkers, maxCapacity, pond.PanicHandler(workerPanicHandler)),
```
i left comment here
https://github.com/prebid/prebid-server/pull/2732/files#r1269240101
let me know what should we do?

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+
+func NewPriceFloorFetcher(maxWorkers, maxCapacity, cacheCleanUpInt, cacheExpiry int) *PriceFloorFetcher {
+	floorFetcher := PriceFloorFetcher{
+		pool:            pond.New(maxWorkers, maxCapacity, pond.PanicHandler(workerPanicHandler)),
```
This library was being used due to performance issue if the number of urls are large
if we think of hosting company, then work with around 10k to 20k publisher, let say every publisher using dynamic floor and everyone has atleast 1 url, then then we have atleast 10k urls
if we try to spawn go routines every time we have a request then we need to create 10k go routines and again we need to fetch them periodically, so again we have to spawn it 
it might cause perfomance issues, thats why used pool instead of creating a go routine for every instance
This library takes care if pool is not getting fully utilised it may shrink or enlarge based on the load

pm-nikhil-vaidya replied:
Let me know if we still want to keep this or switch to normal go-routine per request

pm-nikhil-vaidya replied:
As decided, we will go with this implementation for now, we will switch to different approach at the time of fetcher 2.0

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+	fetchInProgress map[string]bool // Map of URL with fetch status
+	configReceiver  chan FetchInfo  // Channel which recieves URLs to be fetched
+	done            chan struct{}   // Channel to close fetcher
+	cache           *cache.Cache    // cache
+	cacheExpiry     time.Duration   // cache expiry time
```
Nitpick: I dont think these two comments are needed

pm-nikhil-vaidya replied:
Done

pm-nikhil-vaidya commented on config/account.go:
```diff
@@ -65,19 +65,20 @@ type AccountPriceFloors struct {
 	MaxRule                int               `mapstructure:"max_rules" json:"max_rules"`
 	MaxSchemaDims          int               `mapstructure:"max_schema_dims" json:"max_schema_dims"`
-	Fetch                  AccountFloorFetch `mapstructure:"fetch" json:"fetch"`
+	Fetcher                AccountFloorFetch `mapstructure:"fetch" json:"fetch"`
```
In the PRD, the json mapping has `fetch` name

pm-nikhil-vaidya replied:
In the section 4.1

SyntaxNode commented on config/account.go:
```diff
@@ -58,17 +58,30 @@ type AccountCCPA struct {
+	Enabled       bool   `mapstructure:"enabled" json:"enabled"`
+	URL           string `mapstructure:"url" json:"url"`
+	Timeout       int    `mapstructure:"timeout_ms" json:"timeout_ms"`
+	MaxFileSize   int    `mapstructure:"max_file_size_kb" json:"max_file_size_kb"`
```
Please update the `MaxFileSize` name to include the `kb` / `kilobyte` measurement that's present in the json name.

pm-nikhil-vaidya replied:
Done

SyntaxNode commented on config/account.go:
```diff
@@ -81,6 +94,34 @@ func (pf *AccountPriceFloors) validate(errs []error) []error {
+		errs = append(errs, fmt.Errorf(`account_defaults.price_floors.fetch.timeout_ms should be between 10 to 10,000 miliseconds`))
+	}
+
+	if !(pf.Fetcher.MaxRules >= 0 && pf.Fetcher.MaxRules < math.MaxInt32) {
```
It is not necessary to check `pf.Fetcher.MaxRules < math.MaxInt32` since `MaxRules` (and `MaxFileSize`) are 32 bit integers, this expression must always be true.

pm-nikhil-vaidya replied:
Done

SyntaxNode commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+
+	var maxAge int
+	if maxAgeStr := httpResp.Header.Get("max-age"); maxAgeStr != "" {
+		maxAge, _ = strconv.Atoi(maxAgeStr)
```
Please do not ignore errors. The http header could be malformed.

pm-nikhil-vaidya replied:
Done

SyntaxNode commented on floors/fetcher.go:
```diff
@@ -0,0 +1,294 @@
+	var maxAge int
+	if maxAgeStr := httpResp.Header.Get("max-age"); maxAgeStr != "" {
+		maxAge, _ = strconv.Atoi(maxAgeStr)
+		if maxAge <= configs.Period || maxAge > math.MaxInt32 {
```
The `maxAge > math.MaxInt32` can never be true since `maxAge int` can never be larger than `MaxInt32`.

pm-nikhil-vaidya replied:
Done

pm-nikhil-vaidya commented:
@SyntaxNode
In the last conversation on the design implementation, we have decided that we can implement the normal go routine.
As we are going to modify the fetcher in later code optimisation i.e fetcher 2.0, can we go head with this implementation instead of changing this again. Even if we modify the implementation the code will be short lived so we can do the changes at the time of fetcher 2.0
Let me know your thoughts on this

hhhjort commented:
Please fix the merge conflict.

pm-nikhil-vaidya commented:
Modified the code which addresses below points

1. Switched from go-cache to freecache
2. introduced httpclient and its config instead of using default http client
3. Used time interface from timeutil
4. introduced metrics engine instance in fetcher

pm-nikhil-vaidya commented on floors/fetcher_test.go:
```diff
@@ -829,6 +829,56 @@ func TestFetcherDataNotPresentInCache(t *testing.T) {
+}
+
+func TestFetcherDynamicFetchDisable(t *testing.T) {
+
```
Done

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,301 @@
+	}
+}
+
+func fetchAndValidate(configs config.AccountFloorFetch) (*openrtb_ext.PriceFloorRules, int) {
```
Done

AlexBVolcy commented on config/account.go:
```diff
@@ -81,6 +94,34 @@ func (pf *AccountPriceFloors) validate(errs []error) []error {
+	}
+
+	if !(pf.Fetcher.MaxSchemaDims >= 0 && pf.Fetcher.MaxSchemaDims < 20) {
+		errs = append(errs, fmt.Errorf(`account_defaults.price_floors.fetch.max_schema_dims should not be less than 0 seconds and greater than 20`))
```
Can you add a test to cover this line?

guscarreon commented on exchange/exchange.go:
```diff
@@ -267,7 +269,7 @@ func (e *exchange) HoldAuction(ctx context.Context, r *AuctionRequest, debugLog
 
 	if e.floor.Enabled {
-		floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions)
+		floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions, e.priceFloorFetcher)
```
If `e.floor.Enabled` evaluates to `false`, we don't do price floors anywhere in the code at all, right? It seems that we don't fetch for price floors nor we check for price floors in the bid request itself neither here nor in line 389 below. Therefore, I don't think we need to initialize the fetcher at all if `e.floor.Enabled` is `false`. `e.floor.Enabled` is a host config value that doesn't change from request to request. If `e.floor.Enabled` is false, we could optimize a bit by not initializing the `PriceFloorFetcher` that comes with `Ticker` and channels overhead. Maybe we could check for this value back in `router.go` and pass a `nil` `PriceFloorFetcher` to the `Exchange` if `e.floor.Enabled` is `false`:
```
227
    +   var priceFloorFetcher *PriceFloorFetcher = nil
    +   if cfg.PriceFloors.Enabled {
228         priceFloorFetcher := floors.NewPriceFloorFetcher(cfg.PriceFloors.Fetcher.Worker, cfg.PriceFloors.Fetcher.Capacity,
229             cfg.PriceFloors.Fetcher.CacheSize, floorFechterHttpClient, r.MetricsEngine)
    +   }
230
231     tmaxAdjustments := exchange.ProcessTMaxAdjustments(cfg.TmaxAdjustments)
232     planBuilder := hooks.NewExecutionPlanBuilder(cfg.Hooks, repo)
233     macroReplacer := macros.NewStringIndexBasedReplacer()
234     theExchange := exchange.NewExchange(adapters, cacheClient, cfg, syncersByBidder, r.MetricsEngine, cfg.BidderInfos, gdprPermsBuilder, rateConvertor, categoriesFetcher, adsCertSigner
235     var uuidGenerator uuidutil.UUIDRandomGenerator
router/router.go
```
We could do a `nil` check here instead: 
```
231   func (e *exchange) HoldAuction(ctx context.Context, r *AuctionRequest, debugLog *DebugLog) (*AuctionResponse, error) {
232       if r == nil {
233           return nil, nil
234       }
235 *-- 35 lines: var floorErrs []error------------------------------------------------------------------------------------------------------------------------------------------------------
270  
271 -     if e.floor.Enabled {
    +     if priceFloorFetcher != nil {
272           floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions, e.priceFloorFetcher)
273       }
exchange/exchange.go
```
Thoughts?

pm-nikhil-vaidya replied:
Sure, will modify the code
Also i think we don't need this check again
```go
if priceFloorFetcher != nil {
       floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions, e.priceFloorFetcher)
}
```
if floors enabled is `true`, then we are creating the floor fetcher instance, so we will not get condition where floors enabled is `true` and fetcher instance is `nil`

pm-nikhil-vaidya replied:
Instead for SafeGuard i will add check when we actually use fetcher instance
```
 func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
-       if !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
+       if f == nil || !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
                return nil, openrtb_ext.FetchNone
        }
```

pm-nikhil-vaidya replied:
Also
Instead of writing the below logic in `router.go`
```
 var priceFloorFetcher *PriceFloorFetcher = nil
    +   if cfg.PriceFloors.Enabled {
228         priceFloorFetcher := floors.NewPriceFloorFetcher(cfg.PriceFloors.Fetcher.Worker, cfg.PriceFloors.Fetcher.Capacity,
229             cfg.PriceFloors.Fetcher.CacheSize, floorFechterHttpClient, r.MetricsEngine)
    +   }
```
The internal implementation of creating fetcher instance will take care of it
```
-func NewPriceFloorFetcher(maxWorkers, maxCapacity, cacheSize int, httpClient *http.Client, metricEngine metrics.MetricsEngine) *PriceFloorFetcher {
+func NewPriceFloorFetcher(config config.PriceFloors, httpClient *http.Client, metricEngine metrics.MetricsEngine) *PriceFloorFetcher {
+       if !config.Enabled {
+               return nil
+       }
+
        floorFetcher := PriceFloorFetcher{
-               pool:            pond.New(maxWorkers, maxCapacity, pond.PanicHandler(workerPanicHandler)),
+               pool:            pond.New(config.Fetcher.Worker, config.Fetcher.Capacity, pond.PanicHandler(workerPanicHandler)),
                fetchQueue:      make(FetchQueue, 0, 100),
                fetchInProgress: make(map[string]bool),
-               configReceiver:  make(chan FetchInfo, maxCapacity),
+               configReceiver:  make(chan FetchInfo, config.Fetcher.Capacity),
                done:            make(chan struct{}),
-               cache:           freecache.NewCache(cacheSize * 1024 * 1024),
+               cache:           freecache.NewCache(config.Fetcher.CacheSize * 1024 * 1024),
                httpClient:      httpClient,
                time:            &timeutil.RealTime{},
                metricEngine:    metricEngine,
```

guscarreon commented on exchange/exchange.go:
```diff
@@ -175,6 +176,7 @@ func NewExchange(adapters map[openrtb_ext.BidderName]AdaptedBidder, cache prebid
 		requestSplitter:          requestSplitter,
 		macroReplacer:            macroReplacer,
 		floor:                    cfg.PriceFloors,
+		priceFloorFetcher:        priceFloorFetcher,
```
We probably don't need to pass the entire `config.PriceFloors` object to the `exchange struct` given that we only check for the `Enabled` value from here on. Unless we plan to use other fields inside `floor` in the future, can we chage to a boolean?
```
82       macroReplacer            macros.Replacer
83 -     floor                    config.PriceFloors
   +     priceFloorEnabled        bool
84       priceFloorFetcher        floors.FloorFetcher
85   }
exchange/exchange.go
```

pm-nikhil-vaidya replied:
Agree, we don't need entire object, it was passed if we need something from config in the future

For now i don't think we need entire object, i will make the change

guscarreon reviewed (commented):
Left a couple of comments below. Also, after this PR gets merged, would we still need this log entry?
```
if cfg.PriceFloors.Enabled {
    glog.Warning(`cfg.PriceFloors.Enabled will currently not do anything as price floors feature is still under development.`)
}
config/config.go
```

guscarreon commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+	}
+
+	var priceFloors openrtb_ext.PriceFloorRules
+	if err = json.Unmarshal(floorResp, &priceFloors.Data); err != nil {
```
This was hard to spot. Only noticed until I tested locally. We are trying to unmarshal into `&priceFloors` not `&priceFloors.Data`

pm-nikhil-vaidya replied:
Actually, only data object will be hosted not entire floor json
It is mentioned in the PRD

pm-nikhil-vaidya replied:
```
When data is retrieved from a dynamic fetch, it must be assumed to be entirely under the 'data' element of Schema 2.
As described in the signaling section below, fetch data will be merged into the floors request config under 'data' before validating. 
The idea is that the request/storedrequest will define some pieces of the overall floor config
(e.g.floorMin, enabled) and the floor provider supplies the actual floor rules under the 'data' section.
```
From PRD

guscarreon commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+				f.submit(nextFetch.(*FetchInfo))
+			}
+		case <-f.done:
+			glog.Info("Price Floor fetcher terminated")
```
Should we make the `f.Stop()` call here?

pm-nikhil-vaidya replied:
When we close the channel, it will send single on the channel and it will take care to stop the fetcher

guscarreon replied:
Where will the `f.Stop()` call take place?

pm-nikhil-vaidya replied:
Ideally it should be called when we are shutting down the server
Currently we are not stopping the fetcher instance, so it was not being used

pm-nikhil-vaidya replied:
This utility is for graceful shutdown of priceFloor fetcher

guscarreon commented:
For the testing purposes, I wrote the very basic price floor server shown below:
```
var port string = "9090"

func main() {
	http.HandleFunc("/", usersHandleFunc)
	log.Printf("Mock floor bid server starts at: http://localhost:" + port)
	log.Fatal(http.ListenAndServe(":"+port, nil))
}

func usersHandleFunc(w http.ResponseWriter, r *http.Request) {
	mockResponse := getMockResponse()

	w.Write(mockResponse)
	return
}

func getMockResponse() []byte {
	return []byte(`{
  "floormin": 5,
  "floormincur": "USD",
  "skiprate": 1,
  "data": {
    "currency": "USD",
    "skiprate": 1,
    "floorsschemaversion": "1.0",
    "modeltimestamp": 10,
    "modelgroups": [
      {
        "currency": "USD",
        "modelversion": "Version 101",
        "skiprate": 1,
        "schema": {
          "fields": [
            "mediaType",
            "size",
            "domain"
          ],
          "delimiter": "|"
        },
        "values": {
          "banner|300x600|www.website5.com": 15,
          "*|*|*": 25
        }
      }
    ],
    "floorprovider": "appnexus"
  }
}`)
}
```
And found the unmarshalling bug mentioned in a previous comment above but, I believe we need further testing end to end testing in a pod

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+
+	close(f.done)
+	f.pool.Stop()
+	close(f.configReceiver)
```
Close(done) will send signal on the done channel
I will add ticker.stop()

pm-nikhil-vaidya replied:
Added ticker.Stop() while closing fetcher instance

```go
@@ -215,6 +215,7 @@ func (f *PriceFloorFetcher) Fetcher() {
                                f.submit(nextFetch.(*FetchInfo))
                        }
                case <-f.done:
+                       ticker.Stop()
                        glog.Info("Price Floor fetcher terminated")
                        return
                }
```

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+var refetchCheckInterval = 300
+
+type FetchInfo struct {
+	config.AccountFloorFetch
```
For my own understanding, why does this line not have a variable name attached to it?

pm-nikhil-vaidya replied:
This is Struct embedding in Golang
With this we can inherit method which are defined on the internal struct. Also we can directly access the attributes inside internal struct without calling internal struct name explicitly

For more details you can refer
[Embedding](https://go.dev/doc/effective_go#embedding)
https://gobyexample.com/struct-embedding

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+	}
+
+	// Check for floors JSON in cache
+	result, found := f.Get(config.Fetcher.URL)
+	if found {
+		var fetchedFloorData openrtb_ext.PriceFloorRules
+		err := json.Unmarshal(result, &fetchedFloorData)
+		if err != nil || fetchedFloorData.Data == nil {
+			return nil, openrtb_ext.FetchError
+		}
+		return &fetchedFloorData, openrtb_ext.FetchSuccess
+	}
```
I think this function can be re-worked slightly so the Get/Unmarshal is on the same line as the if statement: 


```
	if result, found := f.Get(config.Fetcher.URL); found {
		var fetchedFloorData openrtb_ext.PriceFloorRules
		if err := json.Unmarshal(result, &fetchedFloorData); err != nil || fetchedFloorData.Data == nil {
			return nil, openrtb_ext.FetchError
		}
		return &fetchedFloorData, openrtb_ext.FetchSuccess
	}
```

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+		return nil, 0, errors.New("error while getting response from url : " + err.Error())
+	}
+
+	if httpResp.StatusCode != 200 {
```
Instead of hardcoding `200`, I suggest using `http.StatusOK` instead

pm-nikhil-vaidya replied:
Done

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+
+}
+
+// Stop terminates fetcher service can closes all the channels and destroys worker pool
```
Done

pm-nikhil-vaidya commented on floors/floors.go:
```diff
@@ -135,12 +136,21 @@ func isPriceFloorsEnabledForRequest(bidRequestWrapper *openrtb_ext.RequestWrappe
+	account.PriceFloors.Fetcher.AccountID = account.ID
+	fetchResult, fetchStatus := priceFloorFetcher.Fetch(account.PriceFloors)
+
+	if account.PriceFloors.UseDynamicData && fetchResult != nil && fetchStatus == openrtb_ext.FetchSuccess {
```
I agree it will avoid a extra function call, but for abstraction purpose i kept all the condition check inside the Fetch() call. This will make code more readable and flexible
Let me know your thoughts on this otherwise i will modify it

pm-nikhil-vaidya replied:
Modified the code

pm-nikhil-vaidya commented:
@guscarreon 
I have tested this mechanism end to end
Let me know any issues you encountered while testing end to end if by chance i missed something

AlexBVolcy commented on floors/fetcher.go:
```diff
@@ -0,0 +1,323 @@
+	// Check for floors JSON in cache
+	if result, found := f.Get(config.Fetcher.URL); found {
+		var fetchedFloorData openrtb_ext.PriceFloorRules
+		err := json.Unmarshal(result, &fetchedFloorData)
+		if err != nil || fetchedFloorData.Data == nil {
+			return nil, openrtb_ext.FetchError
```
Can these lines be re-written as: 

```
if err := json.Unmarshal(result, &fetchedFloorData); err != nil || fetchedFloorData.Data == nil {
	return nil, openrtb_ext.FetchError
}
```

pm-nikhil-vaidya replied:
Done

AlexBVolcy commented:
This is looking good to me, but I'm going to wait to see what other comments @guscarreon makes before approving.

pm-nikhil-vaidya commented on floors/floors.go:
```diff
@@ -135,12 +136,26 @@ func isPriceFloorsEnabledForRequest(bidRequestWrapper *openrtb_ext.RequestWrappe
+	var fetchResult *openrtb_ext.PriceFloorRules
+	var fetchStatus string
+	if account.PriceFloors.UseDynamicData {
+		fetchResult, fetchStatus = priceFloorFetcher.Fetch(account.PriceFloors)
```
Fixed

AlexBVolcy reviewed (commented):
Looks good to me! Will wait for @guscarreon before approving.

guscarreon commented on floors/fetcher.go:
```diff
@@ -0,0 +1,321 @@
+	}
+
+	// Send to refetch channel
+	f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
```
Suppose the `config.AccountFloorFetch` comes with a mispelled URL. We would expect `f.fetchFloorRulesFromURL(config)` to always return an error in line 224 below:
```
223 func (f *PriceFloorFetcher) fetchAndValidate(config config.AccountFloorFetch) (*openrtb_ext.PriceFloorRules, int) {
224     floorResp, maxAge, err := f.fetchFloorRulesFromURL(config)
225     if err != nil {
226         glog.Errorf("Error while fetching floor data from URL: %s, reason : %s", config.URL, err.Error())
227         return nil, 0
228     }
floors/fetcher.go
```
Do we want to refetch? Should we not discard this faulty client info altogether?
```
152   func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
153 -     floorData, fetchedMaxAge := f.fetchAndValidate(config)
    +     floorData, fetchedMaxAge, err := f.fetchAndValidate(config)
    +     if isReoccurringError(err) {
    +         return
    +     }
154       if floorData != nil {
155           // Update cache with new floor rules
156           glog.Infof("Updating Value in cache for URL %s", config.URL)
157           cacheExpiry := config.MaxAge
158           if fetchedMaxAge != 0 && fetchedMaxAge > config.Period && fetchedMaxAge < math.MaxInt32 {
159               cacheExpiry = fetchedMaxAge
160           }
161           floorData, err := json.Marshal(floorData)
162           if err != nil {
163               glog.Errorf("Error while marshaling fetched floor data for url %s", config.URL)
164           } else {
165               f.SetWithExpiry(config.URL, floorData, cacheExpiry)
166           }
167       }
168  
169       // Send to refetch channel
170       f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
171  
172   }
floors/fetcher.go
```
The implementation or `isReoccurringError(err)` should be able to tell apart situations where it's not worth trying to fetch again. Thoughts?

pm-nikhil-vaidya replied:
If the URL is invalid then we do not pass it to fetcher as we have check in place,`validator.IsURL(config.Fetcher.URL)`

But if your saying the unreachable for long time, then we can have mechanism to discard the URL, but we don't have any way to find out is that url is temporarily down or permanent

We have to keep track of failures

let  me know your thoughts on this

guscarreon replied:
What do you think about a max number of retries? Maybe that field could be added to the `FetchInfo` struct and made host-configurable.
```
152 - func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
152 + func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch, hostMaxPriceFloorFetchRetries int) {
153       floorData, fetchedMaxAge := f.fetchAndValidate(config)
154       if floorData != nil {
155           // Update cache with new floor rules
156           glog.Infof("Updating Value in cache for URL %s", config.URL)
157           cacheExpiry := config.MaxAge
158           if fetchedMaxAge != 0 {
159               cacheExpiry = fetchedMaxAge
160           }
161           floorData, err := json.Marshal(floorData)
162           if err != nil {
163               glog.Errorf("Error while marshaling fetched floor data for url %s", config.URL)
164           } else {
165               f.SetWithExpiry(config.URL, floorData, cacheExpiry)
166           }
167       }
168  
169       // Send to refetch channel
    +     if config.retryCount < hostMaxPriceFloorFetchRetries { // horrible name LOL
    +         config.retryCount++
170           f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
    +     }
171  
172   }
floors/fetcher.go
```
The reason I believe this is necessary is because `time.Duration(config.Period)` gets added to `time.Now()` in line 170 above making an incorrect URL to throw 404s indefinitely every 5 minutes. What do you think?

pm-nikhil-vaidya replied:
It seems right to me, to discard the invalid URLs
I will add mechanism to discard long failing URLs

pm-nikhil-vaidya replied:
Also what should be default value for retries?

pm-nikhil-vaidya replied:
Default value set to 10

guscarreon commented on floors/fetcher.go:
```diff
@@ -0,0 +1,321 @@
+		glog.Infof("Updating Value in cache for URL %s", config.URL)
+		cacheExpiry := config.MaxAge
+		if fetchedMaxAge != 0 && fetchedMaxAge > config.Period && fetchedMaxAge < math.MaxInt32 {
+			cacheExpiry = fetchedMaxAge
```
Nitpick: I noticed we are comparing `fetchedMaxAge` against `config.Period` twice. The first time happens inside `fetchFloorRulesFromURL`. Should we avoid this second comparison altogether by moving this logic to `fetchFloorRulesFromURL` somehow?
```
251   func (f *PriceFloorFetcher) fetchFloorRulesFromURL(config config.AccountFloorFetch) ([]byte, int, error) {
252 *-- 16 lines: ctx, cancel := context.WithTimeout(context.Background(), time.Duration(config.Timeout)*time.Millisecond)------------------------------------------
268  
269 -     var maxAge int
    +     var maxAge int = config.MaxAge
270       if maxAgeStr := httpResp.Header.Get("max-age"); maxAgeStr != "" {
271 -         maxAge, err = strconv.Atoi(maxAgeStr)
    +         headerMaxAge, err = strconv.Atoi(maxAgeStr)
272           if err != nil {
273               glog.Errorf("max-age in header is malformed for url %s", config.URL)
274 -         }
275 -         if maxAge <= config.Period {
    +         } else if headerMaxAge <= config.Period || headerMaxAge >= math.MaxInt32 {
276               glog.Errorf("Invalid max-age = %s provided, value should be valid integer and should be within (%v, %v)", maxAgeStr, config.Period, math.MaxInt32)
277 -         }
    +         } else {
                  maxAge = headerMaxAge
    +         }
278       }
279  
floors/fetcher.go
```

pm-nikhil-vaidya replied:
Done

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,321 @@
+}
+
+func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
+	if f == nil || !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
```
These checks kept for safely using the methods if caller does not handle these cases
Let me know if we still need to remove it

guscarreon commented on config/account.go:
```diff
@@ -82,6 +95,34 @@ func (pf *AccountPriceFloors) validate(errs []error) []error {
+		errs = append(errs, fmt.Errorf(`account_defaults.price_floors.fetch.period_sec should be less than account_defaults.price_floors.fetch.max_age_sec`))
+	}
+
+	if pf.Fetcher.Period < 300 {
```
The code as written, I don't think will send more than one request every 5 minutes for a given heap element if this value is less than 300:
```
209         case <-ticker.C:
210             currentTime := f.time.Now().Unix()
211             for top := f.fetchQueue.Top(); top != nil && top.FetchTime <= currentTime; top = f.fetchQueue.Top() {
212                 nextFetch := heap.Pop(&f.fetchQueue)
213                 f.submit(nextFetch.(*FetchInfo))
214             }
floors/fetcher.go
```
If less than 5 minutes, the element will evaluate to `top.FetchTime <= currentTime` every time the `<-ticker` sends a signal. Do we need this check?

pm-nikhil-vaidya replied:
I did not get it, can you please explain bit more on this

guscarreon replied:
It was better explained in my first comment above. No need to remove the check.

pm-nikhil-vaidya commented on floors/fetcher.go:
```diff
@@ -0,0 +1,321 @@
+			}
+		case <-ticker.C:
+			currentTime := f.time.Now().Unix()
+			for top := f.fetchQueue.Top(); top != nil && top.FetchTime <= currentTime; top = f.fetchQueue.Top() {
```
It just to optimise traversal of heap, at any point of time we don't have much url to refetch so waiting for 5 mins to gather for refetch and also if any URL is down temporally then it will get us some time and avoid failure

guscarreon replied:
> It just to optimise traversal of heap, at any point of time we don't have much url to refetch so waiting for 5 mins to gather for refetch and also if any URL is down temporally then it will get us some time and avoid failure.

Ok. Sounds reasonable.

guscarreon replied:
> If you think this is unnecessary then i will remove it

No, that's fine

guscarreon commented:
Aside from the last comment above, we also need to address the merge conflicts.

pm-nikhil-vaidya commented:
@guscarreon 
Added retry mechanism if URL is url is long dead

Max retries are configurable through config file
```go
type PriceFloorFetcher struct {
	HttpClient HTTPClient `mapstructure:"http_client"`
	CacheSize  int        `mapstructure:"cache_size_mb"`
	Worker     int        `mapstructure:"worker"`
	Capacity   int        `mapstructure:"capacity"`
	MaxRetries int        `mapstructure:"max_retries"`
}
```

This config is consumed when floor fetcher is initialised
```go
floorFetcher := PriceFloorFetcher{
		pool:            pond.New(config.Fetcher.Worker, config.Fetcher.Capacity, pond.PanicHandler(workerPanicHandler)),
		fetchQueue:      make(FetchQueue, 0, 100),
		fetchInProgress: make(map[string]bool),
		configReceiver:  make(chan fetchInfo, config.Fetcher.Capacity),
		done:            make(chan struct{}),
		cache:           freecache.NewCache(config.Fetcher.CacheSize * 1024 * 1024),
		httpClient:      httpClient,
		time:            &timeutil.RealTime{},
		metricEngine:    metricEngine,
		maxRetries:      config.Fetcher.MaxRetries,
	}
```
The worker will try to fetch the floor data from url and if fetch fails then it will increment the retry count in fetchInfo else it will reset the count value
```go
func (f *PriceFloorFetcher) worker(fetchConfig fetchInfo) {
	floorData, fetchedMaxAge := f.fetchAndValidate(fetchConfig.AccountFloorFetch)
	if floorData != nil {
		// Reset retry count when data is successfully fetched
		fetchConfig.retryCount = 0

		// Update cache with new floor rules
		cacheExpiry := fetchConfig.AccountFloorFetch.MaxAge
		if fetchedMaxAge != 0 {
			cacheExpiry = fetchedMaxAge
		}
		floorData, err := json.Marshal(floorData)
		if err != nil {
			glog.Errorf("Error while marshaling fetched floor data for url %s", fetchConfig.AccountFloorFetch.URL)
		} else {
			f.SetWithExpiry(fetchConfig.AccountFloorFetch.URL, floorData, cacheExpiry)
		}
	} else {
		fetchConfig.retryCount++
	}

	// Send to refetch channel
	if fetchConfig.retryCount < f.maxRetries {
		fetchConfig.fetchTime = f.time.Now().Add(time.Duration(fetchConfig.AccountFloorFetch.Period) * time.Second).Unix()
		fetchConfig.refetchRequest = true
		f.configReceiver <- fetchConfig
	}
}
```

pm-nikhil-vaidya commented:
@guscarreon @SyntaxNode @bsardo @hhhjort @AlexBVolcy 

I have addressed all the review comments. Kindly review and let me know if anything needs to be addressed so that we can finalise this implementation.

guscarreon reviewed (commented):
After data was fetched and cached successfully, do we need to refetch? What would be the purpose? To update the floorprice?
Or, should we stop refetching and, if the `cacheExpiry` is met, floor info us simply removed from cache and will be refetched whenever a new request comes in? 
```
155   func (f *PriceFloorFetcher) worker(fetchConfig fetchInfo) {
156       floorData, fetchedMaxAge := f.fetchAndValidate(fetchConfig.AccountFloorFetch)
157       if floorData != nil {
158           // Reset retry count when data is successfully fetched
159           fetchConfig.retryCount = 0
160  
161           // Update cache with new floor rules
162           cacheExpiry := fetchConfig.AccountFloorFetch.MaxAge
163           if fetchedMaxAge != 0 {
164               cacheExpiry = fetchedMaxAge
165           }
166           floorData, err := json.Marshal(floorData)
167           if err != nil {
168               glog.Errorf("Error while marshaling fetched floor data for url %s", fetchConfig.AccountFloorFetch.URL)
169           } else {
170               f.SetWithExpiry(fetchConfig.AccountFloorFetch.URL, floorData, cacheExpiry)
171           }
    +         
    +         return
    +         
172       } else {
173           fetchConfig.retryCount++
174       }
175  
176       // Send to refetch channel
177       if fetchConfig.retryCount < f.maxRetries {
178           fetchConfig.fetchTime = f.time.Now().Add(time.Duration(fetchConfig.AccountFloorFetch.Period) * time.Second).Unix()
179           fetchConfig.refetchRequest = true
180           f.configReceiver <- fetchConfig
181       }
182   }
floors/fetcher.go
```
Thoughts?

guscarreon commented:
> Also what should be default value for retries?

I'm sorry I missed your question. Sure, 10 is fine

pm-nikhil-vaidya commented:
The refetch mechanism is mainly introduced as this hosted floor data will be updated by ML algorithms , ML will constantly learn by applying particular floor value to request and based on bids it will increase or decrease the floor value, so yes we will require refetch mechanism.

pm-nikhil-vaidya commented:
PRD also mentions about refetch mechanism
https://docs.google.com/document/d/1YAt00RfQaU2Ugciqw1LnirLgjyUz9ibfyjfxWUHR_P0/edit#heading=h.yglzagui7gzp

guscarreon reviewed (approved):
Thanks for addressing my feedback @pm-nikhil-vaidya . LGTM

SyntaxNode commented:
Prebid Server 2.0 has been released and Go Module name has changed from `github.com/prebid/prebid-server` to `github.com/prebid/prebid-server/v2`, per Go versioning conventions.

Please merge the `master` branch and update all `prebid-server` import statements to reference the `v2` name change. Thank you.

pm-nikhil-vaidya commented:
@SyntaxNode @guscarreon 

Updated to V2

guscarreon reviewed (approved):
LGTM

AlexBVolcy reviewed (approved):
LGTM
